docs/design-documents/2020-02-Promtail-Push-API.md

```diff
@@ -0,0 +1,124 @@
+# Promtail Push API
+
+- Author: Robert Fratto (@rfratto)
+- Date: Feb 4 2020
+- Status: DRAFT
+
+Despite being an optional piece of software, Promtail provides half the power
+of Loki's story: log transformations, service discovery, metrics from logs,
+and context switching between your existing metrics and logs. Today, Promtail
+can only be operated to consume logs from very specific sources: files, journal,
+or syslog. If users wanted to write custom tooling to ship logs, the tooling
+has to bypass Promtail and push directly to Loki. This can lead users to
+reimplement functionality Promtail already provides, including its error retries
+and batching code.
+
+This document proposes a Push API for Promtail. By being compatible with the
+Loki Push API, the Promtail Push API can allow batches of logs to be processed
+at once for optimizing performance. Matching the Promtail API also allows users
+to transparently switch their push URLs from their existing tooling. Finally,
+a series of alternative solutions will be detailed.
+
+## Configuration
+
+Promtail will have a new target called HTTPTarget, configurable in the
+`scrape_config` array with the following schema:
+
+```yaml
+# Defines an HTTP target, which exposes an endpoint against the Promtail
+# HTTP server to accept log traffic.
+http:
+  # Defines the base URL for the push path, adding a prefix to the
+  # exposed endpoint. The final endpoint path is
+  # <base_url>loki/api/v1/push. If omitted, defaults to /.
+  #
+  # Multiple http targets with the same base_url must not exist.
+  base_url: /
+
+  # Map of labels to add to every log line passed through to the target.
+  labels: {}
+```
+
+### Considerations
+
+Users will be able to define multiple `http` scrape configs, but the base URL
+value must be different for each instance. This allows to cleanly separate
+pipelines through different push endpoints.
+
+Users must also be aware about problems with running Promtail with an HTTP
```

Comment:
The load balancing issue is very annoying and makes scaling difficult (users have to implement sharding or something else themselves).

I'm not sure I like it, but we could always support a simple sharding config where you can specify multiple promql endpoints and we can shard between them via labels. Changing this hash ring in such a primitive mode would require service restarts, though.

Reply:
Yeah, that's a good point. I'm tempted to call this out of scope and handle it later on. IMO, the most flexible thing to do would be to use a DNS lookup to find all of our promtails and use that for sharding; we want to be able to support non-Kubernetes.

Reply:
I'd be afraid to support DNS in sharding because of how ephemeral it may be. You'd still likely run into the same problems as you would with traditional round robin load balancing, albeit less often. That's why I was considering a simple inline config that hardcodes a number of endpoints, like:

```
promtails:
  prom1.<namespace>.svc.cluster.local
  prom2.<namespace>.svc.cluster.local
  prom3.<namespace>.svc.cluster.local
```

Note: I'm definitely not in love with this, but I'm trying to think of an _expedient_ solution here.

Reply:
Owen and I had a quick chat offline and agreed that we shouldn't try to deal with load balancing ourselves right now. There's a few points to be made here: 

1. If you hardcode endpoints, you can't dynamically scale without rolling all your instances. 
2. If you use DNS, the sharding is eventually consistent and might cause improper writes. 
3. If instances rotate, the old instance might still be buffering some streams. If a new instance sends data for new values for streams before the old instance finishes flushing its buffer, you'll get out of order errors.  

This is definitely a problem, but it's really complicated and not worth solving in Promtail right now. The (unfortunate) recommendation will be for people to write their own sharding mechanism in front of Promtail if they need to scale, and just use one Promtail instance otherwise.

Reply:
I think considering it out of scope until you have feedback on your first version is the right way to go, but since we are discussing possibilities, @owen-d has a good idea. Round robin over  configured list is a solid option - it's flexible enough to work, and is simple. 

That said, I always like when things seamlessly integrate with consul :)

Reply:
Not happy with the consensus here. We are going to make it worse and it's already a problem.

Reply:
We may want to revisit the assumption that we _need_ to be able to forward logs to multiple promtails.

Reply:
Technically with the proposed API, you could chain together Promtails by having one Promtail write to another. 

What use case are you thinking about?